### PR TITLE
Implement rand()

### LINKED ISSUE #109: rand()
First of all, thank you, @daanhb and everyone else for making this package - we have been needing something to represent these types of spaces for a while: https://github.com/JuliaPOMDP/POMDPs.jl/issues/301, https://github.com/JuliaReinforcementLearning/CommonRLSpaces.jl/issues/14 (also, it seems quite a challenge to deal with all of the conceptual issues and different use cases people have - I admire your valor in attempting that :smiley: )

One thing that we commonly need in our applications is to get a random element of a set, for example `rand(ProductDomain(1..2, 3..4))`.

Do you think that this package should include `rand`?

On one hand, this does not make a ton of sense because it is not clear what distribution this should be drawn from (uniform would be a natural choice, but may be hard to implement for some sets), but on the other hand, many other sets without distributions support `rand`, i.e. (`rand(Set([1,2]))`, `rand(1..2)`).

If `rand` does not belong in this package, where should it go?

### PR DESCRIPTION
**Pull request by zsunberg**
Fix #109

### CODE REVIEW

**Comment by zsunberg** [comment]
It took me like 3 hours to understand the type system, but I think this whole implementation is going to be < 50 lines when done. :joy:

**Comment by daanhb** [comment]
Yes, that sounds very Julian: the ratio of time it takes to write code versus length of code can be pretty high (especially with lots of types flying around working together in undocumented ways as is the case here...sorry about that).

On the one hand, one expects there to be an implementation of `rand` for some or all of the "basic" types. And then it should all "just work" together. For inspiration on that second part you can look at how `point_in_domain` works. For example, product domains are dealt with using [this line](https://github.com/JuliaApproximation/DomainSets.jl/blob/60bb050ad8e7622611ff4cb412de06a7145b017c/src/generic/lazy.jl#L85):
```julia
point_in_domain(d::CompositeDomain) = toexternalpoint(d, map(point_in_domain, components(d)))
```

The `map` command will apply `point_in_domain` to all `components` of the composite domain. (It is called `factors` only for product domains, `components` is slightly more general). The "toexternalpoint" machinery will translate the internal representation of the element to the external one (of type T). That could mean that a tuple of a Float64 and an SVector{2,Float64} is concatenated into an SVector{3}, in case of a product domain of an interval with a disk, say. That is something `map` would not do on its own. It works like this:

```julia

julia> using DomainSets

julia> using DomainSets: ×

julia> d = (1..2.0) × UnitDisk()
(1.0..2.0) × UnitDisk()

julia> eltype(d)
SVector{3, Float64} (alias for StaticArrays.SArray{Tuple{3}, Float64, 1, 3})

julia> x = point_in_domain(d)
3-element StaticArrays.SVector{3, Float64} with indices SOneTo(3):
 1.5
 0.0
 0.0

julia> DomainSets.tointernalpoint(d, x)
(1.5, [0.0, 0.0])

julia> typeof(ans)
Tuple{Float64, StaticArrays.SVector{2, Float64}}

julia> map(DomainSets.point_in_domain, components(d))
(1.5, [0.0, 0.0])

julia> DomainSets.toexternalpoint(d, ans)
3-element StaticArrays.SVector{3, Float64} with indices SOneTo(3):
 1.5
 0.0
 0.0
```
The internal representation works will with `map`, the external representation doesn't but makes more sense for the user.

Probably rand only makes sense for product domains. It would be difficult to define properly for unions, intersections and differences of domains (the other composite domains), because there is no guarantee about whether or not the composing domains have any overlap - the distribution could be skewed because of that. Same for mapped domains: some kind of random point could be defined but, unless the map is affine, what is its distribution?

File: src/applications/random.jl [hunk 1]
```diff
@@ -0,0 +1,18 @@
+Random.gentype(::Type{<:Domain{T}}) where T = T
+
+# Random.gentype(::Type{<:ProductDomain{T}}) where T = T
+Base.rand(rng::AbstractRNG, s::Random.SamplerTrivial{<:ProductDomain}) = map(i->(rand(rng, i)), factors(s[]))
```
Comment (daanhb) [inline src/applications/random.jl:4]: I think here you'd want to put a `toexternalpoint(d, outcome_of_map)` around it, where `d` is the product domain

**Comment by zsunberg** [comment]
@daanhb I implemented the easy ones. I think this is a good enough minimal implementation to be merged, but I would be willing to take a stab at it if you think others should be included.

**Comment by daanhb** [comment]
> @daanhb I implemented the easy ones. I think this is a good enough minimal implementation to be merged, but I would be willing to take a stab at it if you think others should be included.

Just the easy cases is fine! If it is isn't easy, then perhaps it should not be in this package after all.

File: src/applications/random.jl [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+function Base.rand(rng::AbstractRNG, s::Random.SamplerTrivial{<:Ball})
+    # Technical details: http://extremelearning.com.au/how-to-generate-uniformly-random-points-on-n-spheres-and-n-balls/
+
+    b = s[]
+    
+    # for low-dimensional balls, use rejection sampling - acceptance rate is at least 52%
+    if dimension(b) <= 3
+        while true
+            r = rand(rng, boundingbox(b))
```
Comment (daanhb) [inline src/applications/random.jl:15]: `boundingbox(d)` allocates memory and is not very optimized, perhaps it is better to define `box=boundingbox(b)` once before the while loop?

**Comment by daanhb** [comment]
This looks good. Could you try to add my suggestions in the issues you opened and see if that fixes the broken tests?

File: src/applications/random.jl (hunk 2 above)
Comment (zsunberg) [reply src/applications/random.jl:15]: > boundingbox(d) allocates memory and is not very optimized

allocates memory even with StaticArrays? that is unfortunate.

**Comment by zsunberg** [comment]
Ok, everything is fixed - I think it's ready to go!

**Comment by daanhb** [comment]
I just noticed that
```julia
julia> using DomainSets

julia> rand(ChebyshevInterval())
ERROR: ArgumentError: Sampler for this object is not defined
```
which also causes
```julia
julia> rand(UnitBall())
ERROR: ArgumentError: Sampler for this object is not defined
```

The reason is that `rand` in IntervalSets is defined for `ClosedInterval`. I think it could be defined for the more general `<:TypedEndpointsInterval{:closed,:closed,T}`, and then `rand` would also work for all of our intervals with fixed endpoints.

**Comment by zsunberg** [comment]
FYI @daanhb I fixed the issue you mentioned above at JuliaMath/IntervalSets.jl#116

**Pull request merged**